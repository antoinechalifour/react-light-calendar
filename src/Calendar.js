import React, { Component } from 'react'
import moment from 'moment'
import Header from './Header'
import Body from './Body'

class Calendar extends Component {
  constructor (props) {
    super(props)

    const { date } = this.props

    this.state = {
      currentMonth: date || moment()
    }

    this.handleDateChange = this.handleDateChange.bind(this)
  }

  handleDateChange (nextDate) {
    const { onChange } = this.props
    const { currentMonth } = this.state

    // When the users clicks a date that does not belong
    // to the currently shown month, we need to update
    // the calendar to that month
    if (nextDate.month() !== currentMonth.month()) {
      this.setState({ currentMonth: nextDate })
    }

    onChange(nextDate)
  }

  render () {
    const { currentMonth } = this.state
    const {
      date,
      theme = {},
      activatedIntervals,
      disabledIntervals
    } = this.props
    const {
      // Name of the current month + navigation arrows
      header: headerStyle,

      // Cell containing each day of the week name
      dayName: nameStyle,

      // style for cells that does not belong to the current month
      // (ex: dec. 31st when current month is January)
      outOfMonth: outOfMonthStyle,

      // style for cells that belong the the shown month
      currentMonth: currentMonthStyle,

      // today cell
      today: todayStyle,

      // currently chose cell
      selected: selectedStyle,

      // disabled cells
      disabled: disabledStyle
    } = theme

    return (
      <div>
        <Header
          currentMonth={currentMonth}
          onChange={nextDisplayedMonth => this.setState({ currentMonth: nextDisplayedMonth })}
          style={{ headerStyle }}
        />
        <Body
          active={date}
          currentMonth={currentMonth}
          onChange={this.handleDateChange}
          activatedIntervals={activatedIntervals}
          disabledIntervals={disabledIntervals}
          style={{
            nameStyle,
            outOfMonthStyle,
            currentMonthStyle,
            todayStyle,
            selectedStyle,
            disabledStyle
          }}
        />
      </div>
    )
  }
}

Calendar.propTypes = {
  // The currently selected date
  date: React.PropTypes.object,
  // Called whenever the user changes the date
  onChange: React.PropTypes.func.isRequired,

  // Only one of the two can be provided to the component
  activatedIntervals: React.PropTypes.array,
  disabledIntervals: React.PropTypes.array
}

export default Calendar
