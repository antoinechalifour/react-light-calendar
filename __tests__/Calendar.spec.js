jest.mock('../src/Header', () => 'Header')
jest.mock('../src/Body', () => 'Body')

import React from 'react'
import ReactTestRenderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import moment from 'moment'
import Calendar from '../src/Calendar'

it('Should set the default month to the given date', () => {
  const date = moment('01/03/2016', 'DD/MM/YYYY')
  const result = shallow(
    <Calendar
      date={date}
      onChange={function onChange () {}}
    />
  )

  const componentState = result.state()

  expect(date.isSame(componentState.currentMonth, 'month')).toBe(true)
})

it('Should set the default month to the current one if no date is given', () => {
  const result = shallow(
    <Calendar
      onChange={function onChange () {}}
    />
  )

  const componentState = result.state()
  const currentMonth = moment()

  expect(currentMonth.isSame(componentState.currentMonth, 'month')).toBe(true)
})

it('Should provide style to its children based on the theme', () => {
  const date = moment('01/02/2014', 'DD/MM/YYYY')
  const renderer = ReactTestRenderer.create(
    <Calendar
      date={date}
      onChange={function onChange () {}}
      theme={{
        header: {
          fontFamily: 'cursive',
          fontStyle: 'italic',
          textTransform: 'none',
          color: '#323a45'
        },
        dayName: {
          textTransform: 'uppercase',
          fontFamily: 'cursive',
          fontStyle: 'italic',
          fontWeight: 'normal'
        },
        outOfMonth: {
          color: '#979899',
          fontFamily: 'cursive'
        },
        currentMonth: {
          color: 'green',
          fontFamily: 'cursive'
        },
        today: {
          fontStyle: 'normal',
          fontWeight: 'italic'
        },
        selected: {
          backgroundColor: '#ff003c',
          color: '#fff'
        },
        disabled: {
          color: '#ff003c'
        }
      }}
    />
  )

  expect(renderer.toJSON()).toMatchSnapshot()
})

it('Should change the displayed month when the picked date is not in range', () => {
  const initialDate = moment('01/01/2017', 'DD/MM/YYYY')
  const nextDate = moment('01/02/2017', 'DD/MM/YYYY')
  const onChange = jest.fn()
  const wrapper = shallow(
    <Calendar
      date={initialDate}
      onChange={onChange}
    />
  )

  expect(wrapper.state().currentMonth.isSame(initialDate, 'month')).toBe(true)
  wrapper.instance().handleDateChange(nextDate)
  expect(wrapper.state().currentMonth.isSame(nextDate, 'month')).toBe(true)

  const argument = onChange.mock.calls[0][0]
  expect(argument).toBe(nextDate)
})
