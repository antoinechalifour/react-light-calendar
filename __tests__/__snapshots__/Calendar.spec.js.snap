exports[`test Should provide style to its children based on the theme 1`] = `
<div>
  <Header
    currentMonth={"2014-01-31T23:00:00.000Z"}
    onChange={[Function]}
    style={
      Object {
        "headerStyle": Object {
          "color": "#323a45",
          "fontFamily": "cursive",
          "fontStyle": "italic",
          "textTransform": "none",
        },
      }
    } />
  <Body
    activatedIntervals={undefined}
    active={"2014-01-31T23:00:00.000Z"}
    currentMonth={"2014-01-31T23:00:00.000Z"}
    disabledIntervals={undefined}
    onChange={[Function]}
    style={
      Object {
        "currentMonthStyle": Object {
          "color": "green",
          "fontFamily": "cursive",
        },
        "disabledStyle": Object {
          "color": "#ff003c",
        },
        "nameStyle": Object {
          "fontFamily": "cursive",
          "fontStyle": "italic",
          "fontWeight": "normal",
          "textTransform": "uppercase",
        },
        "outOfMonthStyle": Object {
          "color": "#979899",
          "fontFamily": "cursive",
        },
        "selectedStyle": Object {
          "backgroundColor": "#ff003c",
          "color": "#fff",
        },
        "todayStyle": Object {
          "fontStyle": "normal",
          "fontWeight": "italic",
        },
      }
    } />
</div>
`;
